src/AspNet.Security.OAuth.Apple/AppleAuthenticationOptions.cs

```diff
@@ -85,14 +85,14 @@ public new AppleAuthenticationEvents Events
         public string MetadataEndpoint { get; set; } = AppleAuthenticationDefaults.MetadataEndpoint;
 
         /// <summary>
-        /// Gets or sets an optional delegate to get the raw bytes of the client's private key
-        /// which is passed the value of the <see cref="KeyId"/> property and the <see cref="CancellationToken"/>
+        /// Gets or sets an optional delegate to get the client's private key which is passed
+        /// the value of the <see cref="KeyId"/> property and the <see cref="CancellationToken"/>
         /// associated with the current HTTP request.
         /// </summary>
         /// <remarks>
         /// The private key should be in PKCS #8 (<c>.p8</c>) format.
         /// </remarks>
-        public Func<string, CancellationToken, Task<byte[]>>? PrivateKeyBytes { get; set; }
+        public Func<string, CancellationToken, Task<string>>? PrivateKey { get; set; }
```

Comment:
Personally, I've never been a huge fan of representing crypto material as "simple" strings (unpopular opinion: I liked `SecureString` 😄). Should we instead directly return an `ECDsa` instance here? Or an `AsymmetricAlgorithm` if we want to future-proof that delegate to support other algorithm (we never know, Apple may switch back to the good ol' RSA 😄).

If none of these options work for you, I'd probably return a `ReadOnlyMemory<char>` here.

Reply:
The idea behind this API is/was to just get the raw key material out of the file you get from Apple, but do it in a non-implementation specific way, so you could get it from disk, from Azure Key Vault, from an env var, etc. etc.

The `ECDsa` is then created internally from the PEM returned above.

https://github.com/martincostello/AspNet.Security.OAuth.Providers/blob/077853134729f67606d710998c77c8f5ea032e71/src/AspNet.Security.OAuth.Apple/Internal/DefaultAppleClientSecretGenerator.cs#L96-L125

Originally I did it with `byte[]` as it was the lowest-common-denominator, but that lead to ugly hacks like this to be able to create the algorithm from the certificate file:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/6759e8d32b309b4454e3bd1cae52ad519b0dd59f/src/AspNet.Security.OAuth.Apple/AppleAuthenticationOptionsExtensions.cs#L43-L51

So that's why I wanted to swap it out string so I could just use `ImportFromPem()`. I did try with `ReadOnlySpan<char>`, but it wouldn't let me use that. I didn't think of `ReadOnlyMemory<char>` though, and locally that seems to work OK.

I'd be happy to change it to `ReadOnlyMemory<char>` if it's not "too exotic" 😄.

The only downside is that it's _slightly_ less intuitive to use in the Key Vault use case (call to `AsMemory()` at the end):

```csharp
public static AppleAuthenticationOptions UseAzureKeyVaultSecret(
    this AppleAuthenticationOptions options,
    Func<string, CancellationToken, Task<Response<KeyVaultSecret>>> secretProvider)
{
    options.GenerateClientSecret = true;
    options.PrivateKey = async (keyId, cancellationToken) =>
    {
        var secret = await secretProvider(keyId, cancellationToken);
        return secret.Value.Value.AsMemory();
    };

    return options;
}
```

Reply:
> The idea behind this API is/was to just get the raw key material out of the file you get from Apple, but do it in a non-implementation specific way, so you could get it from disk, from Azure Key Vault, from an env var, etc. etc.

Wouldn't moving the responsibility to create the `ECDsa` to the delegate implementer achieve the same goal?

With your example, it would likely look like something like this:

```csharp
public static AppleAuthenticationOptions UseAzureKeyVaultSecret(
    this AppleAuthenticationOptions options,
    Func<string, CancellationToken, Task<Response<KeyVaultSecret>>> secretProvider)
{
    options.GenerateClientSecret = true;
    options.PrivateKey = async (keyId, cancellationToken) =>
    {
        var secret = await secretProvider(keyId, cancellationToken);
        var algorithm = ECDsa.Create();
        algorithm.ImportFromPem(secret.Value.Value);
        return algorithm;
    };

    return options;
}
```

Reply:
Makes sense, let's opt for `ReadOnlyMemory<char>` then 😄 

Since you mentioned Azure Key Vault, it's interesting to note that since we're forced to have the private key material, we can't support scenarios where the key material is stored in a HSM and can't be exported (they call that "hardware keys"). But you're likely right, it's an advanced scenario.